frontend-ssr/components/notifications-tooltip/NotificationsTooltip.component.tsx

```diff
@@ -0,0 +1,221 @@
+  markNotificationAsUnread,
+  markAllAsRead,
+} from '../../redux/user/user.actions';
+import UserService from '../../services/User.service';
+import { timeAgo } from '../../services/Utils';
+
+import Scroll from '../Scroll';
+import LoadingNotification from './loading-notification/LoadingNotification.component';
+import { RootState } from '~/redux/root.reducer';
+import SweetAlertService from '~/services/sweet-alert/SweetAlert.service';
+
+import styles from './NotificationsTooltip.module.scss';
+
+interface State {
+  loading: boolean;
+  isOpen: boolean;
+}
+
+const PRELOAD_DISTANCE = 50;
+
+class NotificationsTooltip extends React.Component<ConnectedProps<typeof connector>, State> {
+  toggleRef: React.RefObject<HTMLDivElement>;
+
+  hiddenRef: React.RefObject<HTMLDivElement>
+
+  scrollRef: React.RefObject<Scroll>
+
+  private observer: IntersectionObserver;
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      loading: false,
+      /** at this point isOpen is true so we can see the notifications when we enter the page */
+      isOpen: true,
+    };
+
+    this.toggleRef = React.createRef();
+    this.hiddenRef = React.createRef();
+    this.scrollRef = React.createRef();
+  }
+
+  componentDidMount() {
+    const { user } = this.props;
+
+    if (!user.end) {
+      this.initIntersectionObserver();
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    const { user } = this.props;
+
+    if (prevProps.user.end !== user.end && user.end) {
+      this.observer.disconnect();
+    }
+
+    if (prevProps.user.end !== user.end && !user.end) {
+      this.initIntersectionObserver();
+    }
+  }
+
+  componentWillUnmount() {
+    const { user } = this.props;
+
+    if (!user.end) {
+      this.observer.disconnect();
+    }
+  }
+
+  initIntersectionObserver = () => {
+    const options = {
+      threshold: 0.3,
+    };
+    this.observer = new IntersectionObserver(this.loadMore, options);
+    this.observer.observe(this.hiddenRef.current);
+  }
+
+  loadMore = (entries) => {
+    entries.forEach((entry) => {
+      if (entry.isIntersecting) {
+        this.loadNextPage();
+      }
+    });
+  }
+
+  loadNextPage = async () => {
+    const { user, dispatch } = this.props;
+    this.setState({ loading: true });
+
+    try {
+      const newNotifications = await UserService.getNotifications(user.page);
+      dispatch(loadNotificationsSuccess(newNotifications));
+      this.scrollRef.current.refresh();
+    } catch (err) {
+      SweetAlertService.toast({ type: 'error', text: err });
+    } finally {
+      this.setState({ loading: false });
+    }
+  };
+
+  markAsRead = async (id) => {
+    const { notifications } = this.props.user;
+    const { dispatch } = this.props;
+
+    if (notifications.list.find((n) => n._id === id).isUnread) {
+      try {
+        await UserService.markAsRead(id);
+        dispatch(markNotificationAsRead(id));
+      } catch (err) {
+        dispatch(markNotificationAsUnread(id));
```

Comment:
Great great job for marking the notification back to `Unread` if the request fails 🎉🎉🎉